pkg/scheduler/internal/cache/node_tree.go

```diff
@@ -41,6 +42,7 @@ type nodeTree struct {
 type nodeArray struct {
 	nodes     []string
 	lastIndex int
+	exhausted bool
```

Comment:
I feel think we don't need private fields (`index` as well as `exhausted`) anymore to record the "state", as the listNode() can serve as a stateless function.

`nodeTree` can be simplified a lot, maybe a simple `map[string][]string` is enough. We just iterator the `nodeTree` like the Python `zip()` function like this: https://play.golang.org/p/g4kmgbFp2_X.

With this, if needed, we even can return the nodes on demand. Just like Python `for ... in xrange` vs. `for ... in range`.

Reply:
I like the simplification, but we better don't introduce intermediate lists. We should just be returning a single list. Or a single element, but that's actually the implementation before this PR that we want to get rid of.

Reply:
I was also expecting that all state related to node iteration be removed, but also without intermediate lists, here is an example: https://play.golang.org/p/6Mf9OrMJCnn

Reply:
aren't 2 nested loops, like proposed in this PR, minus the state, more readable?

Reply:
I feel that both just as readable, note that we only need the `list` and `next` functions, and both seem  straightforward.